homeassistant/components/tankerkoenig/sensor.py

```diff
@@ -0,0 +1,170 @@
+        _LOGGER.error("Error fetching data from tankerkoenig.de:\n%s", data["message"])
+        _LOGGER.error("Could not setup sensors")
+        return
+
+    if len(data["stations"]) <= 0:
+        _LOGGER.warning("Could not find any station in range")
+    else:
+        for station in data["stations"]:
+            master = add_station(station, api_key, fuel_types, entities, master)
+
+    for additional_station_id in additional_stations:
+        try:
+            additional_station_data = pytankerkoenig.getStationData(
+                api_key, additional_station_id
+            )
+        except pytankerkoenig.customException as err:
+            additional_station_data = {"ok": False, "message": err, "exception": True}
+        if not additional_station_data["ok"]:
+            _LOGGER.warning(
+                "Error when adding station %s: %s",
+                additional_station_id,
+                additional_station_data["message"],
+            )
+        else:
+            master = add_station(
+                additional_station_data["station"],
+                api_key,
+                fuel_types,
+                entities,
+                master,
+            )
+
+    async_add_entities(entities, True)
+
+
+def add_station(station, api_key, fuel_types, entities, master):
+    """Add fuel station to the entity list."""
+    _LOGGER.debug(
+        "Add_station called for station: %s and fuel types: %s", station, fuel_types
+    )
+    station_id = station["id"]
+    for fuel in fuel_types:
+        sensor = None
+        if master is None:
+            master = FuelPriceSensorMaster(
+                api_key, fuel, station, f"{NAME}_{station['name']}_{fuel}"
+            )
+            sensor = master
+        else:
+            sensor = FuelPriceSensorSlave(
+                fuel, station, f"{NAME}_{station['name']}_{fuel}"
+            )
+            master.add_slave(station_id, sensor)
+        entities.append(sensor)
+    return master
+
+
+class FuelPriceSensorMaster(FuelPriceSensorBase):
+    """Master sensor for tankerkoenig prices. This sensor gets the updated prices and forwards them to any slaves."""
+
+    def __init__(self, api_key, fuel_type, station, name=NAME):
+        """Initialize the class."""
+        super().__init__(fuel_type, station, name)
+        self._api_key = api_key
+        self._slaves = {}
+        self._monitored_stations = [self._station_id]
+
+    @property
+    def should_poll(self):
+        """Poll regularly for the standalone sensor."""
+        return True
+
+    def add_slave(self, station_id, slave):
+        """Add an additional slave sensor, that needs to be updated together with this one."""
+        if station_id not in self._slaves:
+            self._slaves[station_id] = [slave]
+        else:
+            self._slaves[station_id].append(slave)
+        self._monitored_stations = list(set(self._monitored_stations) | {station_id})
+
+    async def async_update(self):
```

Comment:
Why is this `async_update`? Nothing is `await`ed.

Reply:
The example I based my code on was using async calls, and thus I kept everything async. Is this wasting resources, or just a matter of taste?

Reply:
You have to change this. Mixing async/sync may cause blocking of our async loop, thus the whole system is affected.

A general rule of thumb:

- If library is async -> integration is async
- If library is sync -> integration is sync